Kafka NullPointerException on fetch

Howdy!

Since we upgraded our Kafka cluster to 3.5.1 (from 2.7.0) we've frequently been seeing errors like this:
```
[2023-08-09 15:02:04,003] ERROR [KafkaApi-4] Unexpected error handling request RequestHeader(apiKey=FETCH, apiVersion=15, clientId=kgo, correlationId=29, headerVersion=2) -- FetchRequestData(clusterId
=null, replicaId=-1, replicaState=ReplicaState(replicaId=-1, replicaEpoch=-1), maxWaitMs=5000, minBytes=1, maxBytes=52428800, isolationLevel=0, sessionId=606347500, sessionEpoch=29, topics=[], forgottenTopicsData=[ForgottenTopic(topic='', topicId=AAAAAAAAAAAAAAAAAAAAAA, partitions=[0])], rackId='') with context RequestContext(header=RequestHeader(apiKey=FETCH, apiVersion=15, clientId=kgo, correlationId=29, headerVersion=2), connectionId='10.99.108.16:9192-10.88.7.32:63300-535', clientAddress=/10.88.7.32, principal=User:ANONYMOUS, listenerName=ListenerName(INSIDE), securityProtocol=PLAINTEXT, clientInformation=ClientInformation(softwareName=unknown, softwareVersion=unknown), fromPrivilegedListener=true, principalSerde=Optional[org.apache.kafka.common.security.authenticator.DefaultKafkaPrincipalBuilder@7efc7797]) (kafka.server.KafkaApis)
java.lang.NullPointerException
```

I'm not at all sure this is franz-go's fault, but given that it's complaining about handling a request I thought that just maybe the client is doing something wrong.

I'm still working on getting the relevant debug logs, but this just happened again, and our entire kafka cluster crashed.

Here's a sample of the errors we got from kafka in case they're interesting:
```
[2023-08-09 18:40:01,468] ERROR [ReplicaFetcher replicaId=4, leaderId=6, fetcherId=8] Error for partition xxx-0>
org.apache.kafka.common.errors.KafkaStorageException: Disk error when trying to access log file on the disk.
java.io.IOException: Connection to 6 was disconnected before the response was read
        at org.apache.kafka.clients.NetworkClientUtils.sendAndReceive(NetworkClientUtils.java:99)
        at kafka.server.BrokerBlockingSender.sendRequest(BrokerBlockingSender.scala:113)
        at kafka.server.RemoteLeaderEndPoint.fetch(RemoteLeaderEndPoint.scala:79)
        at kafka.server.AbstractFetcherThread.processFetchRequest(AbstractFetcherThread.scala:316)
        at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3(AbstractFetcherThread.scala:130)
        at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3$adapted(AbstractFetcherThread.scala:129)
        at scala.Option.foreach(Option.scala:437)
        at kafka.server.AbstractFetcherThread.maybeFetch(AbstractFetcherThread.scala:129)
        at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:112)
        at kafka.server.ReplicaFetcherThread.doWork(ReplicaFetcherThread.scala:98)
        at org.apache.kafka.server.util.ShutdownableThread.run(ShutdownableThread.java:127)
```
and
```
[2023-08-09 18:48:27,710] ERROR Closing socket for 10.3.108.16:9192-10.2.22.42:48218-9 because of error (kafka.network.Processor)
org.apache.kafka.common.errors.InvalidRequestException: Error getting request for apiKey: PRODUCE, apiVersion: 9, connectionId: 10.3.108.16:9192-10.2.22.42:48218-9, listenerName: ListenerName(INSIDE), principal:User:ANONYMOUS
Caused by: java.lang.IllegalArgumentException: Varint is too long, the most significant bit in the 5th byte is set, converted value: b645ea89
        at org.apache.kafka.common.utils.ByteUtils.illegalVarintException(ByteUtils.java:453)
        at org.apache.kafka.common.utils.ByteUtils.readUnsignedVarint(ByteUtils.java:160)
        at org.apache.kafka.common.protocol.ByteBufferAccessor.readUnsignedVarint(ByteBufferAccessor.java:70)
        at org.apache.kafka.common.message.ProduceRequestData$TopicProduceData.read(ProduceRequestData.java:545)
        at org.apache.kafka.common.message.ProduceRequestData$TopicProduceData.<init>(ProduceRequestData.java:464)
        at org.apache.kafka.common.message.ProduceRequestData.read(ProduceRequestData.java:172)
        at org.apache.kafka.common.message.ProduceRequestData.<init>(ProduceRequestData.java:114)
        at org.apache.kafka.common.requests.ProduceRequest.parse(ProduceRequest.java:256)
        at org.apache.kafka.common.requests.AbstractRequest.doParseRequest(AbstractRequest.java:171)
        at org.apache.kafka.common.requests.AbstractRequest.parseRequest(AbstractRequest.java:165)
        at org.apache.kafka.common.requests.RequestContext.parseRequest(RequestContext.java:95)
        at kafka.network.RequestChannel$Request.<init>(RequestChannel.scala:102)
        at kafka.network.Processor.$anonfun$processCompletedReceives$1(SocketServer.scala:1148)
        at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:647)
        at kafka.network.Processor.processCompletedReceives(SocketServer.scala:1126)
        at kafka.network.Processor.run(SocketServer.scala:1012)
        at java.base/java.lang.Thread.run(Thread.java:833)
```
and
```
java.lang.NullPointerException: Cannot invoke "String.hashCode()" because the return value of "kafka.server.CachedPartition.topic()" is null
```

--- a/pkg/kgo/metadata.go
+++ b/pkg/kgo/metadata.go
@@ -807,6 +807,18 @@ func (cl *Client) mergeTopicPartitions(
 			)
 		}
 
+		if !isProduce {
+			var noID [16]byte
+			if newTP.cursor.topicID == noID && oldTP.cursor.topicID != noID {
+				cl.cfg.logger.Log(LogLevelWarn, "metadata update is missing the topic ID when we previously had one, ignoring update",
+					"topic", topic,
+					"partition", part,
+				)
+				retryWhy.add(topic, int32(part), errMissingTopicID)
+				continue
+			}
+		}
+
 		// If the tp data is the same, we simply copy over the records
 		// and cursor pointers.
 		//
@@ -866,7 +878,10 @@ func (cl *Client) mergeTopicPartitions(
 	}
 }
 
-var errEpochRewind = errors.New("epoch rewind")
+var (
+	errEpochRewind    = errors.New("epoch rewind")
+	errMissingTopicID = errors.New("missing topic ID")
+)
 
 type multiUpdateWhy map[kerrOrString]map[string]map[int32]struct{}
 

--- a/pkg/kgo/source.go
+++ b/pkg/kgo/source.go
@@ -1871,7 +1871,7 @@ func (f *fetchRequest) adjustPreferringLag() {
 
 func (*fetchRequest) Key() int16 { return 1 }
 func (f *fetchRequest) MaxVersion() int16 {
-	if f.disableIDs {
+	if f.disableIDs || f.session.disableIDs {
 		return 12
 	}
 	return 15
@@ -1905,7 +1905,7 @@ func (f *fetchRequest) AppendTo(dst []byte) []byte {
 		partitions := f.usedOffsets[topic]
 
 		var reqTopic *kmsg.FetchRequestTopic
-		sessionTopic := f.session.lookupTopic(topic)
+		sessionTopic := f.session.lookupTopic(topic, f.topic2id)
 
 		var usedTopic map[int32]struct{}
 		if sessionUsed != nil {
@@ -1972,6 +1972,20 @@ func (f *fetchRequest) AppendTo(dst []byte) []byte {
 			}
 			if len(partitions) == 0 {
 				delete(f.session.used, topic)
+				id := f.session.t2id[topic]
+				delete(f.session.t2id, topic)
+				// If we deleted a topic that was missing an ID, then we clear the
+				// previous disableIDs state and potentially reenable it.
+				var noID [16]byte
+				if id == noID {
+					f.session.disableIDs = false
+					for _, id := range f.session.t2id {
+						if id == noID {
+							f.session.disableIDs = true
+							break
+						}
+					}
+				}
 			}
 		}
 	}
@@ -1996,13 +2010,17 @@ type fetchSession struct {
 	epoch int32
 
 	used map[string]map[int32]fetchSessionOffsetEpoch // what we have in the session so far
+	t2id map[string][16]byte
 
-	killed bool // if we cannot use a session anymore
+	disableIDs bool // if anything in t2id has no ID
+	killed     bool // if we cannot use a session anymore
 }
 
 func (s *fetchSession) kill() {
 	s.epoch = -1
 	s.used = nil
+	s.t2id = nil
+	s.disableIDs = false
 	s.killed = true
 }
 
@@ -2015,6 +2033,8 @@ func (s *fetchSession) reset() {
 	}
 	s.epoch = 0
 	s.used = nil
+	s.t2id = nil
+	s.disableIDs = false
 }
 
 // bumpEpoch bumps the epoch and saves the session id.
@@ -2035,17 +2055,23 @@ func (s *fetchSession) bumpEpoch(id int32) {
 	s.id = id
 }
 
-func (s *fetchSession) lookupTopic(topic string) fetchSessionTopic {
+func (s *fetchSession) lookupTopic(topic string, t2id map[string][16]byte) fetchSessionTopic {
 	if s.killed {
 		return nil
 	}
 	if s.used == nil {
 		s.used = make(map[string]map[int32]fetchSessionOffsetEpoch)
+		s.t2id = make(map[string][16]byte)
 	}
 	t := s.used[topic]
 	if t == nil {
 		t = make(map[int32]fetchSessionOffsetEpoch)
 		s.used[topic] = t
+		id := t2id[topic]
+		s.t2id[topic] = id
+		if id == ([16]byte{}) {
+			s.disableIDs = true
+		}
 	}
 	return t
 }

